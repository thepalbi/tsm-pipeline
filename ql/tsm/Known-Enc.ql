//import javascript
import semmle.javascript.security.dataflow.BrokenCryptoAlgorithmCustomizations
import semmle.javascript.security.dataflow.BuildArtifactLeakCustomizations
import semmle.javascript.security.dataflow.CleartextLoggingCustomizations
import semmle.javascript.security.dataflow.CleartextStorageCustomizations
import semmle.javascript.security.dataflow.ClientSideUrlRedirectCustomizations
import semmle.javascript.security.dataflow.CodeInjectionCustomizations
import semmle.javascript.security.dataflow.CommandInjectionCustomizations
import semmle.javascript.security.dataflow.ConditionalBypassCustomizations
import semmle.javascript.security.dataflow.CorsMisconfigurationForCredentialsCustomizations
import semmle.javascript.security.dataflow.DeepObjectResourceExhaustionCustomizations
import semmle.javascript.security.dataflow.DifferentKindsComparisonBypassCustomizations
import semmle.javascript.security.dataflow.ExternalAPIUsedWithUntrustedDataCustomizations
import semmle.javascript.security.dataflow.FileAccessToHttpCustomizations
import semmle.javascript.security.dataflow.HardcodedCredentialsCustomizations
import semmle.javascript.security.dataflow.HardcodedDataInterpretedAsCodeCustomizations
import semmle.javascript.security.dataflow.HttpToFileAccessCustomizations
import semmle.javascript.security.dataflow.ImproperCodeSanitizationCustomizations
import semmle.javascript.security.dataflow.IncompleteHtmlAttributeSanitizationCustomizations
import semmle.javascript.security.dataflow.IndirectCommandInjectionCustomizations
import semmle.javascript.security.dataflow.InsecureDownloadCustomizations
import semmle.javascript.security.dataflow.InsecureRandomnessCustomizations
import semmle.javascript.security.dataflow.InsufficientPasswordHashCustomizations
import semmle.javascript.security.dataflow.LogInjection
import semmle.javascript.security.dataflow.LoopBoundInjectionCustomizations
import semmle.javascript.security.dataflow.NosqlInjectionCustomizations
import semmle.javascript.security.dataflow.PostMessageStarCustomizations
import semmle.javascript.security.dataflow.PrototypePollutingAssignmentCustomizations
import semmle.javascript.security.dataflow.PrototypePollutionCustomizations
import semmle.javascript.security.dataflow.RegExpInjectionCustomizations
import semmle.javascript.security.dataflow.RemotePropertyInjectionCustomizations
import semmle.javascript.security.dataflow.RequestForgeryCustomizations
import semmle.javascript.security.dataflow.ServerSideUrlRedirectCustomizations
import semmle.javascript.security.dataflow.ShellCommandInjectionFromEnvironmentCustomizations
import semmle.javascript.security.dataflow.SqlInjectionCustomizations
import semmle.javascript.security.dataflow.StackTraceExposureCustomizations
import semmle.javascript.security.dataflow.TaintedFormatStringCustomizations
import semmle.javascript.security.dataflow.TaintedPathCustomizations
import semmle.javascript.security.dataflow.TemplateObjectInjectionCustomizations
import semmle.javascript.security.dataflow.TypeConfusionThroughParameterTamperingCustomizations
import semmle.javascript.security.dataflow.UnsafeDeserializationCustomizations
import semmle.javascript.security.dataflow.UnsafeDynamicMethodAccessCustomizations
import semmle.javascript.security.dataflow.UnsafeJQueryPluginCustomizations
import semmle.javascript.security.dataflow.UnsafeShellCommandConstructionCustomizations
import semmle.javascript.security.dataflow.UnvalidatedDynamicMethodCallCustomizations
import semmle.javascript.security.dataflow.XmlBombCustomizations
import semmle.javascript.security.dataflow.XpathInjectionCustomizations
import semmle.javascript.security.dataflow.Xss
import semmle.javascript.security.dataflow.XxeCustomizations
import semmle.javascript.security.dataflow.ZipSlipCustomizations

predicate isKnownSink(DataFlow::Node nd) {
  nd instanceof BrokenCryptoAlgorithm::Sink
  or
  nd instanceof BuildArtifactLeak::Sink
  or
  nd instanceof CleartextLogging::Sink
  or
  nd instanceof CleartextStorage::Sink
  or
  nd instanceof ClientSideUrlRedirect::Sink
  or
  nd instanceof CodeInjection::Sink
  or
  nd instanceof CommandInjection::Sink
  or
  nd instanceof ConditionalBypass::Sink
  or
  nd instanceof CorsMisconfigurationForCredentials::Sink
  or
  nd instanceof DeepObjectResourceExhaustion::Sink
  or
  nd instanceof DifferentKindsComparisonBypass::Sink
  or
  nd instanceof FileAccessToHttp::Sink
  or
  nd instanceof HardcodedCredentials::Sink
  or
  nd instanceof HardcodedDataInterpretedAsCode::Sink
  or
  nd instanceof HttpToFileAccess::Sink
  or
  nd instanceof ImproperCodeSanitization::Sink
  or
  nd instanceof IncompleteHtmlAttributeSanitization::Sink
  or
  nd instanceof IndirectCommandInjection::Sink
  or
  nd instanceof InsecureDownload::Sink
  or
  nd instanceof InsecureRandomness::Sink
  or
  nd instanceof InsufficientPasswordHash::Sink
  or
  nd instanceof LogInjection::Sink
  or
  nd instanceof LoopBoundInjection::Sink
  or
  nd instanceof NosqlInjection::Sink
  or
  nd instanceof PostMessageStar::Sink
  or
  nd instanceof PrototypePollutingAssignment::Sink
  or
  nd instanceof PrototypePollution::Sink
  or
  nd instanceof RegExpInjection::Sink
  or
  nd instanceof RemotePropertyInjection::Sink
  or
  nd instanceof RequestForgery::Sink
  or
  nd instanceof ServerSideUrlRedirect::Sink
  or
  nd instanceof ShellCommandInjectionFromEnvironment::Sink
  or
  nd instanceof SqlInjection::Sink
  or
  nd instanceof StackTraceExposure::Sink
  or
  nd instanceof TaintedFormatString::Sink
  or
  nd instanceof TaintedPath::Sink
  or
  nd instanceof TemplateObjectInjection::Sink
  or
  nd instanceof TypeConfusionThroughParameterTampering::Sink
  or
  nd instanceof UnsafeDeserialization::Sink
  or
  nd instanceof UnsafeDynamicMethodAccess::Sink
  or
  nd instanceof UnsafeJQueryPlugin::Sink
  or
  nd instanceof UnsafeShellCommandConstruction::Sink
  or
  nd instanceof UnvalidatedDynamicMethodCall::Sink
  or
  nd instanceof XmlBomb::Sink
  or
  nd instanceof XpathInjection::Sink
  or
  nd instanceof Xxe::Sink
  or
  nd instanceof ZipSlip::Sink
  or
  exists(DataFlow::PropWrite pw | isKnownSink(pw.getBase()) | nd = pw.getRhs())
}

// external predicate knownSink(string nd, string url, string q, string rep);
// string buildURL(DataFlow::Node nd) {
//     result =
//       "file://" + nd.getFile().getAbsolutePath() + ":" + nd.getStartLine() + ":" +
//         nd.getStartColumn() + ":" + nd.getEndLine() + ":" + nd.getEndColumn()
//   }
//   string buildURLFromLocation(Location nd) {
//     result =
//       "file://" + nd.getFile().getAbsolutePath() + ":" + nd.getStartLine() + ":" +
//         nd.getStartColumn() + ":" + nd.getEndLine() + ":" + nd.getEndColumn()
//   }
//   string buildURLFromTwoLocations(Location nd1, Location nd2) {
//     result =
//       "file://" + nd1.getFile().getAbsolutePath() + ":" + nd1.getStartLine() + ":" +
//                   nd1.getStartColumn() + ":" + nd2.getEndLine() + ":" + nd2.getEndColumn()
//   }
// DataFlow::Node fromUrl(string url) {
//     exists( DataFlow::Node nd|
//         url = buildURL(nd)
//         and result = nd
//     )
// }
// from string nm, string nurl,  string repr, DataFlow::Node nd, string ls,  string lf
//     where knownSink(nm, nurl, "SqlInjectionWorse", repr) and nd = fromUrl(nurl) and
//     ls = buildURLFromLocation(nd.getEnclosingExpr().getEnclosingStmt().getLocation())
//     and
//     lf = buildURLFromTwoLocations(nd.getContainer().getLocation(),
//                                   nd.getEnclosingExpr().getEnclosingStmt().getLocation())
// select nd, ls,  lf, repr
import tsm.CanonicalReps

from
  DataFlow::Node nd, File f, int startLine, int endLine, int startColumn, int endColumn, string rep,
  int stmtStartLine, int stmtEndLine, int stmtStartColumn, int stmtEndColumn, int funcStartLine,
  int funcStartColumn
where
  isKnownSink(nd) and
  rep = getconcatrep(nd, true) and
  nd.hasLocationInfo(f.getAbsolutePath(), startLine, startColumn, endLine, endColumn) and
  nd.getEnclosingExpr()
      .getEnclosingStmt()
      .getLocation()
      .hasLocationInfo(f.getAbsolutePath(), stmtStartLine, stmtStartColumn, stmtEndLine,
        stmtEndColumn) and
  nd.getContainer()
      .getLocation()
      .hasLocationInfo(f.getAbsolutePath(), funcStartLine, funcStartColumn, _, _)
select f.getAbsolutePath(), f.getRelativePath(), startLine, startColumn, endLine, endColumn, rep,
  stmtStartLine, stmtStartColumn, stmtEndLine, stmtEndColumn, funcStartLine, funcStartColumn
