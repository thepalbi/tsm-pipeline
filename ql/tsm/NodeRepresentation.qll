/**
 * An implementation of access paths for representing events in propogation graphs.
 */

import javascript

/**
 * Holds if `nd` is relevant to program semantics.
 */
predicate isRelevant(DataFlow::Node nd) {
  // exclude externs files (i.e., our manually-written API models) and ambient files (such as
  // TypeScript `.d.ts` files); there is no real data flow going on in those
  not nd.getTopLevel().isExterns() and
  not nd.getTopLevel().isAmbient() and
  nd.getBasicBlock() instanceof ReachableBasicBlock
}

/**
 * Gets the maximum depth of a candidate representation.
 *
 * Increasing this bound will generate more candidate representations, but
 * will generally negatively affect performance. Note that rare candidates are
 * are filtered out above, so increasing the bound beyond a certain threshold may
 * not actually yield new candidates.
 */
private int maxdepth() { result = 5 }

/** Gets a node that the main module of package `pkgName` exports. */
private DataFlow::Node getAnExport(string pkgName) {
  exists(NPMPackage pkg, Module m | pkg.getPackageName() = pkgName and m = pkg.getMainModule() |
    exists(AnalyzedPropertyWrite apw |
      apw.writes(m.(AnalyzedModule).getModuleObject(), "exports", result)
    )
    or
    m.(ES2015Module).exports("default", result.(DataFlow::ValueNode).getAstNode())
  )
}

/** Gets a node that the main module of `pkgName` exports under the name `prop`. */
private DataFlow::Node getAnExport(string pkgName, string prop) {
  exists(NPMPackage pkg, AnalyzedModule m, AnalyzedPropertyWrite apw |
    pkg.getPackageName() = pkgName and
    m = pkg.getMainModule() and
    apw.writes(m.getAnExportsValue(), prop, result)
  )
}

/**
 * Gets a candidate representation of `nd` as a (suffix of an) access path.
 */
string candidateRep(DataFlow::Node nd, int depth, boolean asRhs) {
  // the global object
  isRelevant(nd) and
  nd = DataFlow::globalObjectRef() and
  result = "(global)" and
  depth = 1 and
  asRhs = false
  or
  // package imports/exports
  isRelevant(nd) and
  exists(string pkg |
    nd = DataFlow::moduleImport(pkg) and
    // avoid picking up local imports
    pkg.regexpMatch("[^./].*") and
    asRhs = false
    or
    nd = getAnExport(pkg) and
    asRhs = true
  |
    result = "(root https://www.npmjs.com/package/" + pkg + ")" and
    depth = 1
  )
  or
  // compound representations
  exists(DataFlow::SourceNode base, string step, string baserep |
    (
      (
        baserep = candidateRep(base, depth - 1, false)
        or
        exists(DataFlow::Node rhs |
          baserep = candidateRep(rhs, depth - 1, true) and
          base = rhs.getALocalSource()
        )
      ) and
      // bound maximum depth of candidate representation
      depth <= maxdepth()
      or
      baserep = "*" and
      depth = 1 and
      // avoid creating trivial representations like `(return *)`
      step.regexpMatch("(member|parameter) [^\\d*].*") and
      isRelevant(nd)
    ) and
    result = "(" + step + " " + baserep + ")"
  |
    // members
    exists(DataFlow::PropRef prop |
      prop = base.getAPropertyRead() and
      nd = prop and
      asRhs = false
      or
      prop = base.getAPropertyWrite() and
      nd = prop.(DataFlow::PropWrite).getRhs() and
      asRhs = true
    |
      step = "member " + prop.getPropertyName()
      or
      step = "member *"
    )
    or
    // instances
    (
      nd = base.getAnInstantiation() and
      asRhs = false
      or
      nd = base.(DataFlow::ClassNode).getAnInstanceReference() and
      asRhs = false
    ) and
    step = "instance"
    or
    // parameters
    exists(string p |
      exists(int i |
        nd = base.(DataFlow::FunctionNode).getParameter(i) and
        asRhs = false
        or
        nd = base.getAnInvocation().getArgument(i) and
        asRhs = true
      |
        p = i.toString()
      )
      or
      nd = base.(DataFlow::FunctionNode).getAParameter() and
      p = nd.(DataFlow::ParameterNode).getName() and
      asRhs = false
    |
      step = "parameter " + p
    )
    or
    // return values
    (
      nd = base.(DataFlow::FunctionNode).getAReturn() and
      asRhs = true
      or
      nd = base.getAnInvocation() and
      asRhs = false
    ) and
    step = "return"
  )
  or
  // named exports, which are treated as members of packages
  isRelevant(nd) and
  exists(string pkg, string m, string baserep |
    nd = getAnExport(pkg, m) and
    baserep = "(root https://www.npmjs.com/package/" + pkg + ")" and
    result = "(member " + m + " " + [baserep, "*"] + ")" and
    depth = 2 and
    asRhs = true
  )
  or
  // global variables, which are treated as members of the global object
  isRelevant(nd) and
  exists(string g |
    nd = DataFlow::globalVarRef(g) and
    asRhs = false
    or
    exists(AssignExpr assgn |
      assgn.getLhs() = DataFlow::globalVarRef(g).asExpr() and
      nd = assgn.getRhs().flow()
    ) and
    asRhs = true
  |
    result = "(member " + g + " (global))" and
    depth = 2
  )
  or
  // we ignore `await`
  exists(DataFlow::SourceNode base |
    base.flowsToExpr(nd.asExpr().(AwaitExpr).getOperand()) and
    result = candidateRep(base, depth, asRhs)
  )
}
