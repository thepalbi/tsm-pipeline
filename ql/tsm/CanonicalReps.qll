/**
 * An implementation of canonical representations for sinks that represent a more likely
 * representation and allows for using fewer representations for each candidate sink.
 */

import javascript
import NodeRepresentation

/** Canonical representation prioritized for sinks: method parameters. */
private string regExp1() { result = "\\(parameter \\w+ \\(member (\\w+|\\*) .*\\)\\)" }

/** Representation not considered canonical: parameters of deeply nested members. */
private string regExp2() {
  result =
    ".*\\(parameter \\w+ \\(member (\\*|\\w+) \\((member|parameter|return|instance) .*\\)\\)\\)"
}

/**
 *  Prioritizes patterns like `[member _] parameter x (member fun)`
 *  by giving an additional score
 */
int plusForCanonicalRep(string rep, DataFlow::Node nd, int depth) {
  rep = rep(nd, depth, true) and
  (
    (
      // Preferred repr
      rep.regexpMatch(regExp1()) and result = 100
      or
      // member X (preferred rep)
      rep.regexpMatch("\\(member \\w+ " + regExp1() + "\\)") and
      result = 75
      or
      // member * (preferred rep)
      rep.regexpMatch("\\(member \\* " + regExp1() + "\\)") and
      result = 50
    ) and
    not rep.regexpMatch(regExp2())
    or
    result = 0
  )
}

/**
 * Prefers a small number of member and parameters for canonical representation.
 *
 * To-Do: Maybe is easier to define whole canonicalRep with a regular expression.
 */
int plusForPreferredStructForRep(int cm, int cmw, int cr, int cp, int cpr, int croot) {
  cm in [0 .. maxdepth()] and
  cmw in [0 .. maxdepth()] and
  cr in [0 .. maxdepth()] and
  cp in [0 .. maxdepth()] and
  cpr in [0 .. maxdepth()] and
  croot in [0 .. maxdepth()] and
  (
    cm in [1 .. 2] and
    cp in [0 .. 2] and
    cr in [0 .. 2] and
    cmw in [0 .. 3] and
    cpr = 0 and
    croot in [0 .. 1] and
    cm + cp + cr in [2 .. 4] and
    result = 20 - 2 * (cp - 1) - 2 * ((cm - cmw) - 1) - cr - 10 * croot
    or
    result = 0
  )
}

/**
 *  For sinks prioritizes patterns like `(parameter x (member fun) ...)`.
 *  by penalizing more occurences of (pamerameter|member)
 */
predicate isRepWithScore(string rep, DataFlow::Node sink, int depth, boolean asRhs, int score) {
  rep = rep(sink, depth, asRhs) and
  exists(int cm, int cmw, int cr, int cp, int cpr, int croot |
    cm = count(rep.indexOf("member")) and
    cmw = count(rep.indexOf("member *")) and
    cr = count(rep.indexOf("return")) and
    cp = count(rep.indexOf("parameter")) and
    cpr = count(rep.indexOf("parameter -1")) and
    croot = count(rep.indexOf("(root ")) and
    (
      asRhs = true and
      // Penalizes the receivers against members and roots
      score =
        max(plusForCanonicalRep(rep, sink, depth)) +
          max(plusForPreferredStructForRep(cm, cmw, cr, cp, cpr, croot) - (2 * cpr + 2 * cmw))
      or
      asRhs = false and
      // Penalizes the receivers against members and roots
      score = cm * 4 + cr * 3 + cp * 5 - (cpr * 2 + cmw * 2)
    )
  )
}

/**
 * Returns one canonical representation for a node.
 * For sinks it prioritizes patterns like `(parameter x (member fun))`
 * and the use of external functions, penalizes the receiver as parameter.
 */
string chooseBestRep(DataFlow::Node sink, boolean asRhs) {
  result =
    max(string rep, int depth, int score |
      isRepWithScore(rep, sink, depth, asRhs, score)
    |
      rep order by score, depth, rep
      /*desc, depth desc, rep*/
    )
}
