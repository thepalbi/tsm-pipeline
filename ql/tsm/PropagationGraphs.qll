/**
 * An implementation of propagation graphs for JavaScript, inspired by
 * [Seldon](https://dl.acm.org/doi/10.1145/3314221.3314648).
 *
 * The nodes of the propagation graph (also called "events") are simply normal data-flow nodes
 * (`DataFlow::Node`). In addition to normal data flow edges, we add steps from properties to their
 * containing object and from function-call arguments to returns. Additionally, certain control
 * dependencies are turned into data dependencies by representing checks on variables by
 * propagation-graph edges. For details, see predicate `step` below.
 *
 * We use a simple access-path language to represent events; see `NodeRepresentation.qll`.
 *
 * Unlike Seldon, we do not use a points-to analysis to model inter-procedural flow, but instead use
 * the type-tracking functionality from the CodeQL standard library; see predicate `triple`
 * in `Triples.qll`.
 */

import javascript
import NodeRepresentation

/**
 * The name of an npm package that should be considered when building the propagation graph.
 * The filter is applied to sink and source candidates.
 *
 * To customize, implement concrete subclasses of this class.
 *
 * For example, to only consider `mongodb` in for sink candidates you can use
 *
 * ```ql
 * class MongoDbIsInteresting extends InterestingPackageForSinks {
 *   MongoDbIsInteresting() { this = "mongodb" }
 * }
 * ```
 *
 * To do the same with sources:
 * ```ql
 * class MongoDbIsInteresting extends InterestingPackageForSources {
 *   MongoDbIsInteresting() { this = "mongodb" }
 * }
 * ```
 *
 *
 *
 * To consider all imports interesting, both for sorces and sinks use
 *
 * ```ql
 * class AllPackagesAreInteresting extends InterestingPackageForSinks, InterestingPackageForSources {
 *   AllPackagesAreInteresting() { exists(API::moduleImport(this)) }
 * }
 * ```
 */
abstract class InterestingPackageForSinks extends string {
  bindingset[this]
  InterestingPackageForSinks() { any() }
}

abstract class InterestingPackageForSources extends string {
  bindingset[this]
  InterestingPackageForSources() { any() }
}

/**
 * Allows to include additional source/sinks/sanitizers as candidates
 *
 * For example, to add additional sinks
 *
 * ```ql
 * class MoreSinks extends SinkCandidate {
 *    MoreSinks() { this = any(API::moduleImport(_)).getAnArgument() }
 * }
 * ```
 */
abstract class AdditionalSourceCandidate extends DataFlow::Node { }

abstract class AdditionalSinkCandidate extends DataFlow::Node { }

abstract class AdditionalSanitizerCandidate extends DataFlow::Node { }

/** Holds if data read from a use of `f` may originate from package `pkg`. */
predicate mayComeFromLibrary(API::Node f, string pkg) {
  // base case: import
  f = API::moduleImport(pkg)
  or
  // covariant recursive cases: members, instances, results, and promise contents
  // of something that comes from a library may themselves come from that library
  exists(API::Node base | mayComeFromLibrary(base, pkg) |
    f = base.getAMember() or
    f = base.getInstance() or
    f = base.getReturn() or
    f = base.getPromised()
  )
  or
  // contravariant recursive case: parameters of something that escapes to a library
  // may come from that library
  exists(API::Node base | mayEscapeToLibrary(base, pkg) | f = base.getAParameter())
}

/**
 * Holds if data written to a definition of `f` may flow to package `pkg`.
 */
predicate mayEscapeToLibrary(API::Node f, string pkg) {
  // covariant recursive case: members, instances, results, and promise contents of something that
  // escapes to a library may themselves escape to that library
  exists(API::Node base | mayEscapeToLibrary(base, pkg) |
    f = base.getAMember() or
    f = base.getInstance() or
    f = base.getReturn() or
    f = base.getPromised()
  )
  or
  // contravariant recursive case: arguments passed to a function
  // that comes from a library may escape to that library
  exists(API::Node base | mayComeFromLibrary(base, pkg) | f = base.getAParameter())
}

/**
 * Holds if `pred` is a call of the form `f(..., x, ...)` and `succ` is a subsequent
 * use of `x` where the result of the call is either known to be true or known to be
 * false.
 */
private predicate guard(DataFlow::CallNode pred, DataFlow::Node succ) {
  exists(ConditionGuardNode g, SsaVariable v |
    g.getTest() = pred.asExpr() and
    pred.getAnArgument().asExpr() = v.getAUse() and
    succ.asExpr() = v.getAUse() and
    g.dominates(succ.getBasicBlock())
  )
}

/**
 * Holds if `pred` -> `succ` is a known flow step for which we have a model.
 */
predicate knownStep(DataFlow::Node pred, DataFlow::Node succ) {
  // exclude known flow/taint step, except for the step from `x` to `x.p` (since otherwise
  // property reads will never be considered sources)
  any(TaintTracking::AdditionalTaintStep s).step(pred, succ) and
  not succ instanceof DataFlow::PropRead
  or
  exists(DataFlow::AdditionalFlowStep s |
    s.step(pred, succ) or
    s.step(pred, succ, _, _) or
    s.loadStep(pred, succ, _) or
    s.storeStep(pred, succ, _) or
    s.loadStoreStep(pred, succ, _)
  )
}

/**
 * Holds if `u` is a candidate for a taint source.
 */
predicate isSourceCandidate(DataFlow::Node u) {
  exists(API::Node nd |
    mayComeFromLibrary(nd, any(InterestingPackageForSources pkg)) and
    not nd = API::moduleImport(_) and
    u = nd.getAnImmediateUse() and
    exists(rep(u, false)) and
    (
      u instanceof DataFlow::CallNode and
      not u = any(Import i).getImportedModuleNode()
      or
      u instanceof DataFlow::ParameterNode
      or
      u instanceof DataFlow::PropRead
    )
  )
  or
  u instanceof AdditionalSourceCandidate
}

/**
 * Holds if `u` is a candidate for a sanitiser.
 */
predicate isSanitizerCandidate(DataFlow::CallNode u) {
  exists(rep(u, false)) and
  not u = any(Import i).getImportedModuleNode()
  or
  u instanceof AdditionalSanitizerCandidate
}

/**
 * Holds if `d` is a candidate for a taint sink.
 */
predicate isSinkCandidate(DataFlow::Node d) {
  exists(API::Node nd |
    mayEscapeToLibrary(nd, any(InterestingPackageForSinks pkg)) and
    d = nd.getARhs() and
    not knownStep(d, _) and
    exists(rep(d, true)) and
    (
      d = any(ReturnStmt ret).getExpr().flow()
      or
      exists(DataFlow::InvokeNode invk |
        d = invk.(DataFlow::MethodCallNode).getReceiver() or
        d = invk.getAnArgument()
      )
      or
      d = any(DataFlow::PropWrite pw).getRhs()
    )
  )
  or
  d instanceof AdditionalSinkCandidate
}

/**
 * Holds if step `pred` -> `succ` should be considered for the propagation graph.
 */
predicate step(DataFlow::Node pred, DataFlow::Node succ) {
  TaintTracking::localTaintStep(pred, succ)
  or
  succ.(DataFlow::SourceNode).hasPropertyWrite(_, pred)
  or
  succ.(DataFlow::CallNode).getAnArgument() = pred
  or
  guard(pred, succ)
}
