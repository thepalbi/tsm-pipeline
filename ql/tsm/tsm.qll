import javascript

import NodeRepresentation
import tsm_repr_pred

module TSM {
    private import TsmRepr

    predicate isSourceCandidate(DataFlow::Node nd) {
          nd instanceof DataFlow::CallNode or
          nd instanceof DataFlow::PropRead or
          nd instanceof DataFlow::ParameterNode
      }
  
      predicate isSanitizerCandidate(DataFlow::Node nd) {
        nd instanceof DataFlow::CallNode
      }
  
      predicate isSinkCandidate(DataFlow::Node nd) {
        (
          exists(DataFlow::InvokeNode invk |
            nd = invk.getAnArgument()
            or
            nd = invk.(DataFlow::MethodCallNode).getReceiver()
          )
          or
          nd = any(DataFlow::PropWrite pw).getRhs()
        )
      }

    string rep(DataFlow::Node node){
        result = candidateRep(node,_, _)
    }

    predicate isSink(DataFlow::Node node, float score){
        isSinkCandidate(node) and
        (exists(rep(node)) and   score = sum(doGetReprScore(rep(node), "snk"))/count(rep(node)) or
        not exists(rep(node)) and score = 0)
    }

    predicate isSource(DataFlow::Node node, float score){
        isSourceCandidate(node) and
        (exists(rep(node)) and   score = sum(doGetReprScore(rep(node), "src"))/count(rep(node)) or
        not exists(rep(node)) and score = 0)
    }

    predicate isSanitizer(DataFlow::Node node, float score){
        isSanitizerCandidate(node) and
        (exists(rep(node)) and
        score = sum(doGetReprScore(rep(node), "san"))/count(rep(node)) or
        not exists(rep(node)) and score = 0)
    }

    float doGetReprScore(string repr, string t){
        result = TsmRepr::getReprScore(repr, t)
   }    
}