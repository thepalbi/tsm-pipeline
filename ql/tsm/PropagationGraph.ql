/**
 * A query for computing triples in propagation graphs.
 */

import PropagationGraphs

class AllPackagesAreInteresting extends InterestingPackageForSinks, InterestingPackageForSources {
  AllPackagesAreInteresting() { exists(API::moduleImport(this)) }
}

from DataFlow::Node src, DataFlow::Node san, DataFlow::Node snk
where triple(src, san, snk)
select rep(src, false), rep(san, false), rep(snk, true)
