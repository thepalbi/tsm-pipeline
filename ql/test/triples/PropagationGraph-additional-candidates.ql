/**
 * A query for computing triples in propagation graphs.
 */

import tsm.PropagationGraphs

class AdditionalSink extends AdditionalSinkCandidate {
  AdditionalSink() {
    this.getFile().getBaseName() = "test.js" and
    this.getStartLine() = 4 and
    this.getStartColumn() = 10
  }
}

class AdditionalSource extends AdditionalSourceCandidate {
  AdditionalSource() {
    this.getFile().getBaseName() = "test.js" and
    this.getStartLine() = 2 and
    this.getStartColumn() = 15
  }
}

from DataFlow::Node src, DataFlow::Node san, DataFlow::Node snk
where triple(src, san, snk)
select src, san, snk
