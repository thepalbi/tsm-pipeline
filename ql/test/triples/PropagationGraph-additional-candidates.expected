| test.js:2:15:2:26 | lib.source() | test.js:3:17:3:37 | lib.san ... ainted) | test.js:4:10:4:18 | sanitized |
